### PR TITLE
Dynamically set replicas for seed-admission-controller deployment.

### PR DESCRIPTION
**Pull request by Gerrit91**
**How to categorize this PR?**

/area control-plane
/kind enhancement

**What this PR does / why we need it**:

In our test setup we can only have a single node seed cluster. However, the seed-admission-controller is deployed with strong pod anti affinity into the seed's garden namespace with a static amount of 3 replicas. This causes the managed resource for this controller to go unhealthy (the desired replica count cannot be reached).

We deal with this problem by manually adding a HPA that reduces the replica count to 1 after every seed reconciliation. This PR aims for a cleaner solution by dynamically detecting the amount of nodes in the seed cluster, reducing the number of replicas for seed clusters with less than 3 nodes.

**Which issue(s) this PR fixes**:
~

**Special notes for your reviewer**:
This is the follow-up of https://github.com/gardener/gardener/pull/3808.

**Release note**:

```feature operator
Seed clusters with less than three nodes are now supported. In earlier versions of Gardener the seed-admission-controller deployment was causing unhealthy managed resources for small seed clusters, preventing seed bootstrapping from succeeding.
```

### CODE REVIEW

**Comment by gardener-robot-ci-1** [comment]
Thank you @Gerrit91 for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

File: pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go [hunk 1]
```diff
@@ -174,16 +181,25 @@ func (g *gardenerSeedAdmissionController) Deploy(ctx context.Context) error {
 			},
 		}
 
-		deployment = &appsv1.Deployment{
+		// if maxUnavailable would not be set, new pods don't come up in small seed clusters
+		// (due to the pod anti affinity new pods are stuck in pending state)
+		maxUnavailable = intstr.FromInt(1)
```
Comment (vpnachev) [inline pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go:186]: Could this be `0` if the number of nodes is grater than the `replicas`?

Comment (Gerrit91) [reply pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go:186]: I think so, yes, but it also enforces a re-distribution of the pods onto other nodes. I guess, when rolling the pods one by one the `maxUnvailable: 1` configuration should not be too big of a problem regarding time of deployment. What do you think? Do you want default settings back for seed clusters > 3 nodes?

Comment (vpnachev) [reply pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go:186]: Thanks! I am fine with the current proposal. The other 2 replicas could keep serving the requests.

**Review by vpnachev — COMMENTED** [review]
By the way, there is a PDB that will not work very well when the admission controller is deployed with single replica https://github.com/gardener/gardener/blob/d9ffd58cf1b557926b43a788e6cbfaf2491cad3d/pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go#L262-L274

File: pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go [hunk 2]
```diff
@@ -174,16 +181,25 @@ func (g *gardenerSeedAdmissionController) Deploy(ctx context.Context) error {
...
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      deploymentName,
 				Namespace: g.namespace,
 				Labels:    getLabels(),
 			},
 			Spec: appsv1.DeploymentSpec{
 				RevisionHistoryLimit: pointer.Int32Ptr(1),
-				Replicas:             pointer.Int32Ptr(3),
-				Selector:             &metav1.LabelSelector{MatchLabels: getLabels()},
+				Replicas:             pointer.Int32Ptr(replicas),
```
Comment (rfranzke) [inline pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go]:
```suggestion
				Replicas:             &replicas,
```

? :)

File: pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go [hunk 3]
```diff
@@ -405,6 +421,21 @@ func init() {
 	utilruntime.Must(err)
 }
 
+func (g *gardenerSeedAdmissionController) getReplicas(ctx context.Context) (int32, error) {
+	nodeList := &corev1.NodeList{}
+	err := g.client.List(ctx, nodeList)
+	if err != nil {
+		return 0, err
+	}
```
Comment (rfranzke) [inline pkg/operation/botanist/component/seedadmissioncontroller/seedadmissioncontroller.go]: As you are only interested in the number of items in the list, can you use a `PartialObjectMetadataList` like here:

https://github.com/gardener/gardener/blob/5988d5ea7bda62f66253ad473f154be814a8e5b3/pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go#L154-L169

**Review by rfranzke — COMMENTED** [review]
Thanks @Gerrit91 :) For me it looks good, only https://github.com/gardener/gardener/pull/3811#pullrequestreview-626146023 is open.

**Comment by vpnachev** [comment]
> Thanks @Gerrit91 :) For me it looks good, only [#3811 (review)](https://github.com/gardener/gardener/pull/3811#pullrequestreview-626146023) is open.

I think it is safe to switch from `pdb.spec.minAvailable=1` to `pdb.spec.maxUnavailable=1`.

**Comment by Gerrit91** [comment]
Thanks to both of you for your reviews. I am not so experienced with PDBs and therefore I am not really sure if this configuration is perfect. Switched to `pdb.spec.maxUnavailable=1` as proposed by vpnachev.

**Review by vpnachev — APPROVED** [review]
/lgtm
/reviewed ok-to-test
/merge squash

**Comment by Gerrit91** [comment]
Thanks a lot for merging! Very much appreciated. :)

With metal-stack we're a bit behind (currently on 1.14 release), so I hope we will soon come to the point where I can see this working in our real environments. :)

**Pull request merged**